gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -592,6 +593,41 @@ ChannelCredentials createS2ASecuredChannelCredentials() {
     }
   }
 
+  boolean isMtlsS2AHardBoundTokensEnabled() {
+    if (!useS2A) {
+      // If S2A cannot be used, {@code HardBoundTokenTypes.MTLS_S2A} hard bound tokens should not be
+      // used
+      return false;
+    }
+    if (allowedHardBoundTokenTypes == null
+        || credentials == null
+        || !(credentials instanceof ComputeEngineCredentials)) {
+      return false;
+    }
+    for (HardBoundTokenTypes boundTokenTypes : allowedHardBoundTokenTypes) {
+      if (boundTokenTypes == HardBoundTokenTypes.MTLS_S2A) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  CallCredentials createHardBoundTokensCallCredentials(
+      ComputeEngineCredentials.GoogleAuthTransport googleAuthTransport,
+      ComputeEngineCredentials.BindingEnforcement bindingEnforcement) {
+    ComputeEngineCredentials.Builder credsBuilder =
+        ((ComputeEngineCredentials) credentials).toBuilder();
+    // We only set scopes and HTTP transport factory from the original credentials because
+    // only those are used in gRPC CallCredentials to fetch request metadata.
+    return MoreCallCredentials.from(
+        ComputeEngineCredentials.newBuilder()
+            .setScopes(credsBuilder.getScopes())
+            .setHttpTransportFactory(credsBuilder.getHttpTransportFactory())
+            .setGoogleAuthTransport(googleAuthTransport)
+            .setBindingEnforcement(bindingEnforcement)
```

Comment:
Do we need to check if these are null before setting them in the builder? I know from below that we are hard-coding them to be MTLS and ON, but I'm not sure if this is something that may be optional/ customizable in the future.

Reply:
We should be fine to not perform a null check. In [the auth library](https://github.com/googleapis/google-auth-library-java/blob/107a332eee18390f1b286906090fc7ada4874ad5/oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java#L270-L275) we will only use the value if it is non-null.

> I'm not sure if this is something that may be optional/ customizable in the future.
This code isn't going to change in the future. The way that client libraries will opt out of the hard bound tokens feature is going to be by removing `HardBoundTokenTypes.MTLS_S2A` from the list of `allowedHardBoundTokenTypes`. This function (`createHardBoundTokensCallCredentials`) is only going to be called by the below code when `isMtlsS2AHardBoundTokensEnabled` evaluates to true.